File: Data Structures & Algorithms/daily-temperatures/submission-2.py
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        res = [0] * len(temperatures)
        stack = deque()
        stack.append(0)
        for r in range(1, len(temperatures)):
            while (
                stack and 
                temperatures[r] >  temperatures[stack[-1]]
                ):
                l = stack.pop()
                res[l] = r - l
            
            stack.append(r)
        
        return res

File: Data Structures & Algorithms/count-connected-components/submission-2.java
class Solution {
    public int countComponents(int n, int[][] edges) {
        int res = 0;
        HashMap<Integer, List<Integer>> adjHash = new HashMap<>();
        Set<Integer> visited = new HashSet<>();
        for (int[] edge: edges) {
            adjHash.computeIfAbsent(edge[0], k -> new ArrayList<>()).add(edge[1]);
            adjHash.computeIfAbsent(edge[1], k -> new ArrayList<>()).add(edge[0]);
        }


        for (int i = 0; i < n; i++){

            if (visited.contains(i)){
                continue;
            }
            res ++;
            Deque<Integer> stack = new ArrayDeque<>();
            stack.push(i);
            while (!stack.isEmpty()){
                int n1 = stack.pop();
                if (visited.contains(n1)) continue;
                visited.add(n1);
                if (!adjHash.containsKey(n1)){
                    continue;
                }
                for (int n2: adjHash.get(n1)){
                    stack.push(n2);
                }
            }
        }

        return res;
    }
}
